### PR TITLE
Added Prometheus metrics

### PR DESCRIPTION
**Pull request by wolmi**
This PR adds the user stats that currently publishes using Redis, as an example here is a screenshot:

image

In order to work with this implementation, the [prometheus-client-c](https://github.com/digitalocean/prometheus-client-c) is required and you need to install the libraries from this project to compile. The easy way of installing the Prometheus libraries on Debian/Ubuntu environment is following the next steps:

- Download the two libraries deb package
- Install by executing `apt install <file>`

This PR solves #474

### CODE REVIEW

**Comment by misi** [comment]
Many Thanks! 
Great feature!! 
Can you please add a flag in config to enable not only the compile option could enable and disable it. Many Thanks! 
I will merge it in 4.5.2.x

**Comment by misi** [comment]
@wolmi Thanks again!

**Comment by wolmi** [comment]
@misi Please review and let me know if something more is needed

**Comment by cameron-elliott** [comment]
This is a very cool feature, 
but having some experience with both Prometheus and Coturn/redis-statsdb,
I would like to ask some questions, and maybe make some suggestions.

I hope these come across as constructive questions and suggestions, 
that is my intent.


## Background Redis pub-sub interface

redis-statsdb uses Redis pub-sub to 'push' stats metrics to Redis 
channels when specific events occur.

For redis-statsdb, it is your responsibility to be listening on the Redis server when
the metrics are sent by Turnserver to Redis, otherwise the metrics will be lost.

## Background Prometheus HTTP interface

Prometheus is an Http server/endpoint which is polled by clients at whatever frequency
they desire.

With Prometheus metrics, a particular unique metric is usually visible forever (via HTTP) after
it becomes avaible. A particular unique metric generally does not disappear.
This is important, because you have no control over the polling frequency of the clients
polling a Prometheus metric provider. 




##   Please make any monotonic values into Counters

Please consider changing all values that do not decrease (ie: bytes-sent, bytes-received)
 into "counters" and not "gauges".

*Counters have a very powerful constraint, they do not decrease in value.*
This is not immediatly obvious, but this is massively important when
sub-sampling or viewing zoomed-out time periods with tools like Grafana.
If you view a 24 hours of 1 seconds samples (86,400) samples of a Gauge,
you have no understanding what happens in time periods that are merged into one row.
If you view a 24 hours of 1 seconds samples (86,400) samples of a Counter,
you can see exactly what positive deltas have occurred for each aggregate time period.

There is a guarantee by definition that what you view over any time period of a Counter
that occurred within the lifetime of the process was a positive monotonic increase.
(If you have ever used an audio waveform editor, and zoomed out much, all you see is
aliasing of the waveform, and you have no idea what is going on without zooming in.
This issue does not occur with Counters, but this exact issue/problem occurs with Gauges)

Finally, and sorry this is long.
*Counters make it practical to create derivative metrics from raw monotonic values,
ie bytes-per-hour-sent, where these types of derived metrics often cannot be created
from Gauge values in tools like Grafana, because of their non-monotonic nature.*

*Counters make it easy to compute rates of change and have programmatic thresholds
in tools like Grafana that trigger alerts or errors.
For example you can compute the following on a Counter:
bytes-per-minute=(counter2-counter1)/(time2-time1)

So, your monitoring system can issue an alert when bytes-per-minute or 
errors-per-minute or packets-lost-per-minute is too high or too low.
These semantics and meanings for rate-of-change derivates on Counters
are much better defined than they are for Gauges.



## Unique metrics once they appear in the metric-namespace, should not disappear

Just want to double check, once you create and report on a unique metric, 
it will not ever disappear from the metric-list as long as the turnserver process
is still alive, correct?
If metrics disappear, this would be a very unusual operating mode for a 
Prometheus HTTP endpoint.
*Some Prometheus observers will see a very different world than
others depending on their polling frequency.*

**Comment by wolmi** [comment]
Very valuable feedback thanks @cameron-elliott 
I'll review and change the metrics accordingly.

**Comment by cameron-elliott** [comment]
I am glad that was useful, @wolmi, and thank you for considering my feedback.

**Comment by cameron-elliott** [comment]
This PR does bring up bring to the forefront another issue, though. @misi @wolmi @bradleythughes 

This really needs @misi's opinion or other key maintainer's support to consider changing.


## TL;DR:
*The use of near-sequentially assigned session_ids (000000000000000001) by `turnserver` means
systems tracking bandwidth using Redis or this PR may lose rx/tx data because of key collisions, if not very very very carefully designed.*

*A uniform random distribution for the session_id would nearly eliminate this risk for systems tracking bandwidth usage from redis.statsdb or this PR*

The issues occur because of session_ids (000000000000000001) when `turnserver` restarts, and the high likelihood of session_ids collision/reuse when running multiple `turnserver`s in large TURN networks.


Example scenario in typical use with this PR.
The following is based upon the current near-sequential session_id assignment.
`turnserver` runs and handles 100 sessions, stats are collected by Prometheus using the user/realm/session_id values as above.
`turnserver` is stopped and restarted, and 100 more sessions occur.
Let's say 50 of the session_id values are re-used in the 2nd run of `turnserver`,
in this case, 50 different sessions values would be erased and overwritten in the Prometheus
database because of the re-use of the session_id values.

*Given this PR and the typical use of Prometheus metrics, the lack of uniform random session_ids would mean 50 of the original first run sessions metrics would be clobbered and over-written because of session_id collisions, value and tracking of 50 session's rx/tx metrics disappear! poof!*

It may help looking at the Redis monitor capture to understand the assignment of session_id numbers. It shows the near-sequential assignment of session_id numbers.
[Sample Redis monitor capture](https://gist.github.com/cameron-elliott/268ec94ededa125388c72623e06b013e)


The session id may have been randomly assigned in older versions of Coturn:
[2013 snippet of redis-stats output](https://groups.google.com/d/msg/turn-server-project-rfc5766-turn-server/Tx34jSnQM-0/Wp-vt7pZsZMJ)

*This issue has also come up for me building a multi-turnserver network of Coturn instances, of course carefully designed software can overcome the fact that session_id is reused often and near-sequentially assigned, but if the tuple/key of (user/realm/session_id) becomes effectively unique in the future (it is not today with 4.5.1.1), then tracking aggregate bandwidth usage across multiple `turnserver` instances can become much easier to implement.*

I hope this is considered a useful and valuable, it sure takes lots of time to write this stuff up in a way that exposes the possible issues. lol.

**Comment by misi** [comment]
@cameron-elliott , @cameron-elliott unfortunately I am busy with other things but I will get back to you..

**Comment by cameron-elliott** [comment]
Please take all the time in the world, this is a no rush issue. @misi

**Comment by cameron-elliott** [comment]
If there is consensus that having uniform random session ids (possibly with some date bits) is useful,
I could submit a PR to do this.

**Comment by rgarrigue** [comment]
Just a word to say, peoples (well, I to begin with ;) ) are waiting for this one !

**Comment by wolmi** [comment]
I've just changed to use counters instead of gauge.

Ready to review!

**Comment by wolmi** [comment]
@misi let me know if something more is needed before merging 😄

**Comment by misi** [comment]
I want to include in the next release but needs some more time to review and test it. Many thanks and I will get back to you ASAP.

**Comment by misi** [comment]
> This PR does bring up bring to the forefront another issue, though. @misi @wolmi @bradleythughes
> 
> This really needs @misi's opinion or other key maintainer's support to consider changing.
> ## TL;DR:
> 
> _The use of near-sequentially assigned session_ids (000000000000000001) by `turnserver` means systems tracking bandwidth using Redis or this PR may lose rx/tx data because of key collisions, if not very very very carefully designed._
> 
> This PR exposes such instances, which can occur when `turnserver` is restarted or multiple `turnserver` instances data is collected into a single Prometheus instance using this PR.
> 
> (This is not a criticism of this PR or `turnserver` or the current method of assigning session_ids,
> it is really just an observation of a possible improvement to robustness and reliability)
> 
> The issues occur because of session_ids are assigned near-sequentially from (000000000000000001) when `turnserver` restarts.
> 
> _A uniform random distribution for the session_id would nearly eliminate this risk for systems tracking bandwidth usage from redis.statsdb or this PR_
> ### Example scenario in typical use with this PR.
> 
> The following is based upon the current near-sequential session_id assignment.
> `turnserver` runs and handles 100 sessions, stats are collected by Prometheus using the user/realm/session_id values as above.
> `turnserver` is stopped and restarted, and 100 more sessions occur.
> Let's say 50 of the session_id values are re-used in the 2nd run of `turnserver`,
> in this case, 50 different sessions values would be erased and overwritten in the Prometheus
> database because of the re-use of the session_id values.
> 
> _Given this PR and the typical use of Prometheus metrics, the lack of uniform random session_ids would mean 50 of the original first run sessions metrics would be clobbered and over-written because of session_id collisions, value and tracking of 50 session's rx/tx metrics disappear! poof!_
> ### 2nd Example Scenario in use with this PR
> 
> Setup two `turnserver` instances, and have values from both instances collected into a single Prometheus instance. Because the key/tuple of (user/realm/session_id) is not unique across two `turnserver` instances, many sessions will clobber/overwrite values collected from the other `turnserver`. (A uniform random session_id would make risk of this near zero)
> ## Sample log of session_id numbers
> 
> It may help looking at the Redis monitor capture to understand the assignment of session_id numbers. It shows the near-sequential assignment of session_id numbers.
> [Sample Redis monitor capture](https://gist.github.com/cameron-elliott/268ec94ededa125388c72623e06b013e)
> ## Maybe older versions of `turnserver` used random assignment?
> 
> The session id may have been randomly assigned in older versions of Coturn:
> [2013 snippet of redis-stats output](https://groups.google.com/d/msg/turn-server-project-rfc5766-turn-server/Tx34jSnQM-0/Wp-vt7pZsZMJ)
> ## My experience with this issue in multi-turnserver network
> 
> _This issue has also come up for me building a multi-turnserver network of Coturn instances, of course carefully designed software can overcome the fact that session_id is reused often and near-sequentially assigned, but if the tuple/key of (user/realm/session_id) becomes effectively unique in the future (it is not today with 4.5.1.1), then tracking aggregate bandwidth usage across multiple `turnserver` instances can become much easier to implement._
> ## Final Thoughts
> 
> If others found the overall idea useful, I would raise the idea of exposing all 64 bits of the session_id, and maybe a fraction of the bits as the date, and assign most of the bits from a uniform random number.
> 
> I hope this is considered a useful and valuable, it sure takes lots of time to write this stuff up in a way that exposes the possible issues. lol.

AFAIK 3 things gives the global unique id: the timestamp, and the logger source ip address, and TURN session ID 
These 3 gives the unique key value that is globally unique. 

We can improve it, but please consider backward compatibility, so please create a new  option for it. e.g. globa-unique-session-id or similar.

**Comment by nazar-pc** [comment]
Does this provide metrics per-user? If so that is excessively verbose in most cases and will inflate number of unique labels to the point of being hard to use at scale.

**Comment by wolmi** [comment]
I can confirm the sessions_id is hard to use, my first approach was to create a Nodejs based exporter that takes metrics from Redis and exposes the /metrics endpoint. The problem was when Prometheus scrapes all the metrics the high amount of session_id affects Prometheus performances when running queries from Grafana to simply get the total amount of new sessions.

**Comment by nazar-pc** [comment]
I'd prefer `/metrics` to expose just total bitrate, number of sessions, CPU/memory usage and similar general metrics with probably per-protocol granularity, but definitely not per-user/session.

**Comment by wolmi** [comment]
Yes, but as @misi has said it should be better for an improvement with configurable flags when the basic functionality was merged.

At the moment the goal is only to replicat the current metrics on Redis but not adding extra metrics or complexity to the configuration.

**Comment by cameron-elliott** [comment]
### External Trun session runner -> Prometheus metrics on jitter, latency, loss

As a note, I have been running a program which continuously runs Turn sessions using turnutils_uclient in order to monitor the health of a Turn server. It has been running for over a year, and it helps me get notified when turn servers are acting flaky but not 100% down, which happens with VPSs sometimes.

It might someday be mature enough to propose adding to Coturn, but I don't think it is there yet.

If you would like to check it out, you can find it here:

[Limeturnmonitor](https://github.com/cameron-elliott/limeturnmonitor)

I will be writing an article on it's benefits and uses.

But, for me it provides 3x Prometheus remote monitoring points for my Coturn/turnserver which allows Prometheus exposure of jitter/latency/loss in a Prometheus format.

This is something the turnserver-internal /metrics cannot really provide. (regular sampling of jitter/latency/loss)

If anyone finds this useful, I will be happy to help them use it,
and maybe someday I could propose a PR for it into Coturn.  Who knows?

Thanks, Cameron

**Comment by wolmi** [comment]
@cameron-elliott Very interesting how you achieve that, if you don't mind I'll use your project as a reference for a new PR to include extra metrics.

**Comment by cameron-elliott** [comment]
@wolmi what do you have in mind? Are you planning a PR to pull the program into Coturn? or something different?  I am planning to do a video tutorial, so that may help people when I get that done.

**Comment by misi** [comment]
I started to review. 
I changed to opt-in. So disabled it by default. 
I adjusted mand page.
Fixed a missing ifdef error of default config.
I saw that the dependeny libs are not part of the debian or ubuntu distribution. :( :(

**Comment by wolmi** [comment]
@cameron-elliott the idea is look at the metrics you are taking from the session on the client and see how they are generated in turnserver and try to add more logic to expose that extra metrics.

@misi I see you have opened an issue on the prometheus-client repo from digitalocean about the official debían package, let me know if I can help.

File: src/apps/relay/ns_ioalib_engine_impl.c [hunk 1]
```diff
@@ -3650,6 +3664,25 @@ void turn_report_allocation_delete(void *a)
...
+#endif
+#if !defined(TURN_NO_PROMETHEUS)
+				{
+					if(ss->realm_options.name[0]){
+						// Set prometheus del metric and update status
+						prom_del_status(ss->realm_options.name, (const char*)ss->username, (unsigned long long)ss->id, (const char *)"deleted");
+
+						// Set prometheus total traffic metrics
+						prom_set_total_traffic(ss->realm_options.name, (const char*)ss->username, (unsigned long long)ss->id, (unsigned long)(ss->t_received_packets), (unsigned long)(ss->t_received_bytes), (unsigned long)(ss->t_sent_packets), (unsigned long)(ss->t_sent_bytes), false);
+						prom_set_total_traffic(ss->realm_options.name, (const char*)ss->username, (unsigned long long)ss->id, (unsigned long)(ss->t_peer_received_packets), (unsigned long)(ss->t_peer_received_bytes), (unsigned long)(ss->t_peer_sent_packets), (unsigned long)(ss->t_peer_sent_bytes), false);
```
Comment (jlao) [inline src/apps/relay/ns_ioalib_engine_impl.c:3676]: Should the last parameter here be `true` since it is a "peer" counter?

File: src/apps/relay/ns_ioalib_engine_impl.c [hunk 2]
```diff
@@ -3650,6 +3664,25 @@ void turn_report_allocation_delete(void *a)
...
+					} else {
+						// Set prometheus del metric and update status
+						prom_del_status(NULL, (const char*)ss->username, (unsigned long long)ss->id,  (const char *)"deleted");
+
+						// Set prometheus total traffic metrics
+						prom_set_total_traffic(ss->realm_options.name, (const char*)ss->username, (unsigned long long)ss->id, (unsigned long)(ss->t_received_packets), (unsigned long)(ss->t_received_bytes), (unsigned long)(ss->t_sent_packets), (unsigned long)(ss->t_sent_bytes), true);
+						prom_set_total_traffic(ss->realm_options.name, (const char*)ss->username, (unsigned long long)ss->id, (unsigned long)(ss->t_peer_received_packets), (unsigned long)(ss->t_peer_received_bytes), (unsigned long)(ss->t_peer_sent_packets), (unsigned long)(ss->t_peer_sent_bytes), true);
```
Comment (jlao) [inline src/apps/relay/ns_ioalib_engine_impl.c:3683]: Should we be passing NULL for the realm here?

File: src/apps/relay/ns_ioalib_engine_impl.c (hunk 1 above)
Comment (wolmi) [reply src/apps/relay/ns_ioalib_engine_impl.c:3676]: Yes it is correct

File: src/apps/relay/ns_ioalib_engine_impl.c (hunk 2 above)
Comment (wolmi) [reply src/apps/relay/ns_ioalib_engine_impl.c:3683]: Also, yes and fixed with the previous comment on next commit

**Comment by wolmi** [comment]
I've changed the default port for the exporter to avoid problems with current allocated ports as described on the Prometheus wiki https://github.com/prometheus/prometheus/wiki/Default-port-allocations, also, adjusted documentation.

**Comment by misi** [comment]
Few comments:
* By default it should be disabled to lower the security risk
* There is a missing ifdef around the default config
* I think the listening port should be configurablet
* run `make-man.sh` to update the man pages and check the linting in both man and readme
* check the linting in --help config of the turn server
* check the linting files you have changed

**Comment by kolajide-proximie** [comment]
Hi guys, I have just pulled the latest code to test out the prometheus endpoint? Prometheus seems to have started according to the output from the app but its not listening on port 9641. Is there anything I need to do on the turnserver.conf to get this to work. I'll post my output here below:


./bin/turnserver -c /etc/turnserver.conf &
[1] 7926
root@ip-172-31-73-62:/usr/local/src/coturn# 0: log file opened: /var/log/turn_7926_2020-08-03.log
0: Listener address to use: 0.0.0.0
0: log file opened: /var/log/turnserver_2020-08-03.log
0: 
RFC 3489/5389/5766/5780/6062/6156 STUN/TURN Server
Version Coturn-4.5.1.3 'dan Eider'
0: 
Max number of open files/sockets allowed for this process: 1048576
0: 
Due to the open files/sockets limitation,
max supported number of TURN Sessions possible is: 524000 (approximately)
0: 

==== Show him the instruments, Practical Frost: ====

0: TLS supported
0: DTLS supported
0: DTLS 1.2 supported
0: TURN/STUN ALPN supported
0: Third-party authorization (oAuth) supported
0: GCM (AEAD) supported
0: OpenSSL compile-time version: OpenSSL 1.1.1  11 Sep 2018 (0x1010100f)
0: 
0: SQLite supported, default database location is /usr/local/var/db/turndb
0: Redis supported
0: PostgreSQL supported
0: MySQL supported
0: MongoDB is not supported
0: 
0: Default Net Engine version: 3 (UDP thread per CPU core)

=====================================================

0: Domain name: 
0: Default realm: *****.******.******
0: ERROR: 
CONFIG ERROR: Empty cli-password, and so telnet cli interface is disabled! Please set a non empty cli-password!
0: SSL23: Certificate file found: /etc/letsencrypt/live/*****.******.******/cert.pem
0: SSL23: Private key file found: /etc/letsencrypt/live/*****.******.******/privkey.pem
0: TLS1.0: Certificate file found: /etc/letsencrypt/live/*****.******.******/cert.pem
0: TLS1.0: Private key file found: /etc/letsencrypt/live/*****.******.******/privkey.pem
0: TLS1.1: Certificate file found: /etc/letsencrypt/live/*****.******.******/cert.pem
0: TLS1.1: Private key file found: /etc/letsencrypt/live/*****.******.******/privkey.pem
0: TLS1.2: Certificate file found: /etc/letsencrypt/live/*****.******.******/cert.pem
0: TLS1.2: Private key file found: /etc/letsencrypt/live/*****.******.******/privkey.pem
0: TLS cipher suite: DEFAULT
0: DTLS: Certificate file found: /etc/letsencrypt/live/*****.******.******/cert.pem
0: DTLS: Private key file found: /etc/letsencrypt/live/*****.******.******/privkey.pem
0: DTLS1.2: Certificate file found: /etc/letsencrypt/live/*****.******.******/cert.pem
0: DTLS1.2: Private key file found: /etc/letsencrypt/live/*****.******.******/privkey.pem
0: DTLS cipher suite: DEFAULT
0: Relay address to use: 0.0.0.0
0: pid file created: /var/run/turnserver.pid
0: IO method (main listener thread): epoll (with changelist)
0: WARNING: I cannot support STUN CHANGE_REQUEST functionality because only one IP address is provided
0: Wait for relay ports initialization...
0:   relay 0.0.0.0 initialization...
0:   relay 0.0.0.0 initialization done
0: Relay ports initialization done
0: IO method (general relay thread): epoll (with changelist)
0: turn server id=0 created
0: IPv4. TLS/SCTP listener opened on : 0.0.0.0:10000
0: IPv4. TLS/TCP listener opened on : 0.0.0.0:10000
0: IPv4. TLS/SCTP listener opened on : 0.0.0.0:5349
0: IPv4. TLS/TCP listener opened on : 0.0.0.0:5349
0: IO method (general relay thread): epoll (with changelist)
0: turn server id=1 created
0: IPv4. DTLS/UDP listener opened on: 0.0.0.0:10000
0: IPv4. TLS/SCTP listener opened on : 0.0.0.0:10000
0: IPv4. TLS/TCP listener opened on : 0.0.0.0:10000
0: IPv4. TLS/SCTP listener opened on : 0.0.0.0:5349
0: IPv4. TLS/TCP listener opened on : 0.0.0.0:5349
0: IPv4. DTLS/UDP listener opened on: 0.0.0.0:5349
0: Total General servers: 2
0: IO method (auth thread): epoll (with changelist)
0: IO method (auth thread): epoll (with changelist)
0: SQLite DB connection success: /usr/local/var/db/turndb
0: IO method (admin thread): epoll (with changelist)
0: 
Prometheus collector started sucessfully.

**Comment by kolajide-proximie** [comment]
Never mind guys, I figured it out, just added 'prometheus' to the config. Cheers

**Comment by misi** [comment]
@kolajide-proximie it is disabled by default to avoid unintentional open ports by default

**Comment by misi** [comment]
@cameronelliott can you write in the INSTALL file  (a new Pull Request) 
few lines about howto setup [Limeturnmonitor](https://github.com/cameron-elliott/limeturnmonitor) with cotrun?  

I would like to link your brilliant project from coturn..

Thanks for your help and cooperation!!

**Comment by cameronelliott** [comment]
@misi  Thank you for the compliment, I very much would like to write an INSTALL PR, and also improve the 
Limeturnmonitor/README. These two tasks are now on my todo list. Please be patient, it may take me a while,
but I will do them.
Cameron

**Comment by ShubhangiSGaikwad** [comment]
> Never mind guys, I figured it out, just added 'prometheus' to the config. Cheers



> Never mind guys, I figured it out, just added 'prometheus' to the config. Cheers

Hello kolajide-proximie , I too tried it but by just adding 'prometheus'  , could you help me on this , nothing appears on localhost:9641/metrics or localhost:9121/metrics

**Comment by rgarrigue** [comment]
When will it be released ? I see no tag / release since the 3 august when it was merged

**Comment by owenmorgan** [comment]
im seeing 

`0: Bad configuration format: prometheus`

on Coturn-4.5.1.3

**Comment by owenmorgan** [comment]
> im seeing
> 
> `0: Bad configuration format: prometheus`
> 
> on Coturn-4.5.1.3

added

`RUN curl -fL -o /tmp/libprom.deb \
         https://github.com/digitalocean/prometheus-client-c/releases/download/v0.1.1/libprom-dev-0.1.1-Linux.deb

RUN curl -fL -o /tmp/libpromhttp.deb \
         https://github.com/digitalocean/prometheus-client-c/releases/download/v0.1.1/libpromhttp-dev-0.1.1-Linux.deb


RUN apt install -y /tmp/libprom.deb
RUN apt install -y /tmp/libpromhttp.deb`

to the Dockerfile and its working

**Comment by os14** [comment]
does it work in 4.5.1.1?

**Comment by corny** [comment]
No, it seems to be released with version 4.5.2
Can't wait to have this feature.
